### PR TITLE
AK+LibGfx: Use SIMD for PNG score calculation

### PR DESCRIPTION
**Pull request by nico**
Produces exactly the same output, but a bit faster.

The speedup is relatively bigger for worse compression:

    image -o sunset_retro.png sunset_retro.bmp --png-compression-level 0
       56.8 ms ±  1.5 ms ->  34.8 ms ± 0.9 ms (38.7% faster)

    image -o sunset_retro.png sunset_retro.bmp --png-compression-level 1
       84.6 ms ±  1.7 ms ->  64.2 ms ± 4.9 ms (24.1% faster)

    image -o sunset_retro.png sunset_retro.bmp --png-compression-level 2
      212.1 ms ±  2.5 ms -> 190.3 ms ± 1.6 ms (10.3% faster)

    image -o sunset_retro.png sunset_retro.bmp --png-compression-level 3
      671.4 ms ± 12.3 ms -> 646.5 ms ± 4.7 ms (3.7% faster)

Compression level 2 is the default, so about a 10% speedup in practice.

For comparison, `sips` needs 49.9 ms ± 3.0 ms to convert sunset_retro.bmp to sunset_retro.png, and judging from the output file size, it uses something similar to our compression level 1. We used to take 1.7x as long as sips, now we take 1.29x as long.

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/PNGWriter.cpp
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/ImageFormats/PNGWriter.cpp:1]: Isn't that just mainly delaying the horizontal add?

File: Userland/Libraries/LibGfx/ImageFormats/PNGWriter.cpp [hunk 1]
```diff
@@ -183,23 +183,25 @@ ErrorOr<void> PNGWriter::add_IDAT_chunk(Gfx::Bitmap const& bitmap, Compress::Zli
...
-            ErrorOr<void> append(u8 byte)
+            ErrorOr<void> append(AK::SIMD::u8x4 simd)
             {
-                TRY(buffer.try_append(byte));
-                sum += static_cast<i8>(byte);
+                TRY(buffer.try_append(simd[0]));
+                TRY(buffer.try_append(simd[1]));
+                TRY(buffer.try_append(simd[2]));
+                if constexpr (include_alpha)
+                    TRY(buffer.try_append(simd[3]));
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/ImageFormats/PNGWriter.cpp:194]: Getting this into one `try_append` is likely also nicer

File: AK/SIMDExtras.h [hunk 2]
```diff
@@ -49,6 +49,12 @@ ALWAYS_INLINE static u32x4 to_u32x4(TSrc v)
     return __builtin_convertvector(v, u32x4);
 }
 
+template<typename TSrc>
+ALWAYS_INLINE static i8x4 to_i8x4(TSrc v)
+{
+    return __builtin_convertvector(v, i8x4);
+}
+
```
Comment (Hendiadyoin1) [inline AK/SIMDExtras.h:57]: #24909 will replace that, with a nicer generic version

File: Userland/Libraries/LibGfx/ImageFormats/PNGWriter.cpp (hunk 1 above)
Comment (nico) [reply Userland/Libraries/LibGfx/ImageFormats/PNGWriter.cpp:194]: I repushed this with slightly more granular commits that maybe make this clearer.

The change is that before we did `sum += (i8)byte;` 3 or 4 times for each pixel's channel, but now we do a single `sum_vec_i32 += (vec_i8)pixel` per pixel (…and then sum up the four channel counts once at the very end).

So it replaces 3 or 4 additions per pixel with a single addition.

File: Userland/Libraries/LibGfx/ImageFormats/PNGWriter.cpp
Comment (nico) [reply Userland/Libraries/LibGfx/ImageFormats/PNGWriter.cpp:1]: I repushed this with slightly more granular commits that maybe make this clearer.

The change is that before we did sum += (i8)byte; 3 or 4 times for each pixel's channel, but now we do a single sum_vec_i32 += (vec_i8)pixel per pixel (…and then sum up the four channel counts once at the very end).

So it replaces 3 or 4 additions per pixel with a single addition.

(I put that into a dedicated commit, and put the no-op moving-around in a separate commit, which hopefully makes this clearer.)

File: Userland/Libraries/LibGfx/ImageFormats/PNGWriter.cpp (hunk 1 above)
Comment (nico) [reply Userland/Libraries/LibGfx/ImageFormats/PNGWriter.cpp:194]: I bet there's a ton more we can do here :)

The actual deflate code can also use some more speed as well.

File: AK/SIMDExtras.h (hunk 2 above)
Comment (nico) [reply AK/SIMDExtras.h:57]: Given time stamps on PRs, I think this one gets right of way :)

**Pull request merged**